pkg/frontend/querymiddleware/astmapper/sharding.go

```diff
@@ -132,17 +136,25 @@ func (summer *shardSummer) MapExpr(expr parser.Expr) (mapped parser.Expr, finish
 		// process in the query-frontend. Since we can't estimate the cardinality, we prefer
 		// to be pessimistic and not parallelize at all the two legs unless we're able to
 		// parallelize all vector selectors in the legs.
-		canShardAllVectorSelectors := func(expr parser.Expr) bool {
-			c := summer.Clone()
-			c.shards = 1
-			m := NewASTExprMapper(c)
+		canShardAllVectorSelectors := func(expr parser.Expr) (can bool) {
+			query := expr.String()
+			// We need to cache the results of this function to avoid processing it again and again
+			// in queries like `a or b or c or d`, which would lead to exponential processing time.
+			if can, ok := summer.canShardAllVectorSelectorsCache[query]; ok {
+				return can
+			}
+			defer func() { summer.canShardAllVectorSelectorsCache[query] = can }()
```

Comment:
how is this caching the result of the function which is `c.stats.GetShardedQueries() == countVectorSelectors(mappedExpr)`

Reply:
`can` is a named return value. So `return` statement assigns the value to it, and then the `defer` statement is executed.

Reply:
ugh, magic, thanks